module-2/routes/users.ts

```diff
@@ -0,0 +1,84 @@
+import express from 'express';
+import { v4 as uuid } from 'uuid';
+import User from '../models/User';
+import { Users } from '../db';
+import schema from '../schema';
+import { filterNotDeleted, getAutoSuggestUsers } from '../utils';
+
+const router = express.Router();
+
+router.get('/', (req, res) => {
+  res.send([...Users.values()].filter(filterNotDeleted));
+});
+
+router.get('/all', (req, res) => {
+  res.send([...Users.values()]);
+});
+
+const DEFAULT_LIMIT = 3;
+router.get('/autosuggest', (req, res) => {
+  const { login = '', limit = DEFAULT_LIMIT } = req.query;
+  res.send(getAutoSuggestUsers(Users)(String(login), Number(limit)));
+});
+
+router.get('/:id', (req, res) => {
+  const user = Users.get(req.params.id);
+  !user?.isDeleted
+    ? res.send(user)
+    : res.status(404).send({ message: 'User not found!' });
+});
+
+router.post('/', (req, res) => {
+  const { login, password, age } = <User>req.body;
+  const { error } = schema.validate({ login, password, age });
+  if (error) {
+    res.status(400).send({ error: error.details.map(detail => detail.message) });
+  } else {
```

Comment:
In Node.js there's this "early return" pattern which is used to check for something important in the beginning of the handler and return/send immediately.
If the check has not passed - everything else in the handler is executed, no "else" case needed.
![image](https://user-images.githubusercontent.com/25229823/91099843-ba057e00-e66c-11ea-9f93-706d7424031d.png)

Reply:
Switched to using early return where applicable, thanks
And have a question regarding usage of this pattern. Is it OK to return from function inside this `if` block in order to stop further function execution? StackOverflow says it's fine 🤨
Ex.:
![image](https://user-images.githubusercontent.com/11832280/91938845-61ce1c00-ecfd-11ea-86e7-3caae09db8eb.png)

Reply:
It's completely OK.
Moreover, it could be used as a part of the logic of your application.
You have early returns, res.send's and "send and return" via `return res.status(...).send(...);`
Which way to go is entirely up to you.